examples/neighbors/plot_lmnn_dim_reduction.py

```diff
@@ -0,0 +1,104 @@
+"""
+===========================================================
+Dimensionality Reduction with Large Margin Nearest Neighbor
+===========================================================
+
+Sample usage of Large Margin Nearest Neighbor for dimensionality reduction.
+
+This example compares different (linear) dimensionality reduction methods
+applied on the Olivetti Faces data set. The data set contains ten different
+images of each of 40 distinct persons. For some subjects, the images were
+taken at different times, varying the lighting, facial expressions (open /
+closed eyes, smiling / not smiling) and facial details (glasses / no glasses).
+Each image of dimensions 64x64 is reduced to a two-dimensional data point.
+
+Principal Component Analysis (PCA) applied to this data identifies the
+combination of attributes (principal components, or directions in the
+feature space) that account for the most variance in the data. Here we
+plot the different samples on the 2 first principal components.
+
+Linear Discriminant Analysis (LDA) tries to identify attributes that
+account for the most variance *between classes*. In particular,
+LDA, in contrast to PCA, is a supervised method, using known class labels.
+
+Large Margin Nearest Neighbor (LMNN) tries to find a feature space such that
+the nearest neighbors classification accuracy is maximized. Like LDA, it is a
+supervised method.
+
+In this example, nearest neighbors classification is used for supervised
+prediction based on the embeddings learned from each of the three
+decomposition approaches, PCA, LDA and LMNN.
+
+One can see that LMNN enforces a clustering of the data that is visually
+meaningful even after the large dimensionality reduction.
+"""
+
+
+# Author: John Chiotellis <johnyc.code@gmail.com>
+# License: BSD 3 clause
+
+import numpy as np
+import matplotlib.pyplot as plt
+from sklearn import datasets
+from sklearn.model_selection import train_test_split
+from sklearn.decomposition import PCA
+from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
+from sklearn.neighbors import LargeMarginNearestNeighbor, KNeighborsClassifier
+
+
+print(__doc__)
+
+n_neighbors = 3
+random_state = 0
+
+# Load Olivetti Faces dataset
+faces = datasets.fetch_olivetti_faces()
+X, y = faces.data, faces.target
+
+# Split into train/test
+X_train, X_test, y_train, y_test = \
+    train_test_split(X, y, test_size=0.5, stratify=y,
+                     random_state=random_state)
+
+dim = len(X[0])
+n_classes = len(np.unique(y))
+
+# Reduce dimension to 2 with PCA
+pca = PCA(n_components=2, random_state=random_state)
+
+# Reduce dimension to 2 with LinearDiscriminantAnalysis
+lda = LinearDiscriminantAnalysis(n_components=2)
+
+# Reduce dimension to 2 with LargeMarginNearestNeighbor
+lmnn = LargeMarginNearestNeighbor(n_neighbors=n_neighbors, n_components=2,
+                                  max_iter=20, verbose=1,
+                                  random_state=random_state)
```

Comment:
The [rendered example](https://129416-843222-gh.circle-artifacts.com/0/doc/auto_examples/neighbors/plot_lmnn_dim_reduction.html) shows a `ConvergenceWarning`, do you think it could be possible to fix it without increasing too much the execution time?

Reply:
When running the file locally, I get the following error that seems to be coming from joblib:

```Traceback (most recent call last):
  File "plot_lmnn_dim_reduction.py", line 53, in <module>
    faces = datasets.fetch_olivetti_faces()
  File "/work/chiotell/projects/scikit-learn/sklearn/utils/validation.py", line 63, in inner_f
    return f(*args, **kwargs)
  File "/work/chiotell/projects/scikit-learn/sklearn/datasets/_olivetti_faces.py", line 120, in fetch_olivetti_faces
    faces = joblib.load(filepath)
  File "/home/johny/soft/miniconda3/envs/scikit/lib/python3.6/site-packages/joblib/numpy_pickle.py", line 585, in load
    obj = _unpickle(fobj, filename, mmap_mode)
  File "/home/johny/soft/miniconda3/envs/scikit/lib/python3.6/site-packages/joblib/numpy_pickle.py", line 504, in _unpickle
    obj = unpickler.load()
  File "/home/johny/soft/miniconda3/envs/scikit/lib/python3.6/pickle.py", line 1050, in load
    dispatch[key[0]](self)
  File "/home/johny/soft/miniconda3/envs/scikit/lib/python3.6/pickle.py", line 1338, in load_global
    klass = self.find_class(module, name)
  File "/home/johny/soft/miniconda3/envs/scikit/lib/python3.6/pickle.py", line 1388, in find_class
    __import__(module, level=0)
ModuleNotFoundError: No module named 'sklearn.externals.joblib.numpy_pickle'
```

I tried with multiple versions of joblib from 0.12.2 to 1.0 and the same error appears. Any ideas?

Reply:
I can't reproduce the error neither with python 3.9 or 3.6 ... have you tried to clean and reinstall? joblib is no longer in externals ...

Reply:
I found it. The problem was with my scikit_learn_data directory. For some reason (probably because it was created by an old `scikit-learn` version) it was causing `pickle` to look for the `sklearn.externals.joblib` instead of `joblib`. I deleted the contents and now it works. I also removed the `ConvergenceWarning` by passing a larger tolerance parameter.